### PR TITLE
Add TextRank visualisation using altair

### LINKED ISSUE #158: Add TextRank visualisations
After talking with @pacoid, I'd like to contribute a visualisation of TextRank to this package.
We already have `BaseTextRank.write_dot` which uses `graphviz`.
The proposed method would be named `BaseTextRank.plot_keyphrases`.

## How it looks

image

```python
import pathlib

import spacy
import pytextrank
import altair as alt
import pandas as pd

nlp = spacy.load("en_core_web_sm")
nlp.add_pipe("textrank", last=True)
text = pathlib.Path("dat/cfc.txt").read_text()
doc = nlp(text)

def plot_keyphrases(doc: spacy.tokens.Doc) -> alt.Chart:
    """Plot profile of a doc's keyphrases' ranks."""
    source = pd.DataFrame([p.__dict__ for p in doc._.phrases]).drop("chunks", axis=1).reset_index()
    c = (
        alt.Chart(source)
        .mark_bar()
        .encode(x="index", y="rank", color="count", tooltip=["text", "rank", "count"])
        .properties(title="Keyphrase profile of the article")
    )
    return c

plot_keyphrases(doc)
```

## Problem

The issue is that this very simple code requires to bring 2 (big) dependencies: `altair` and `pandas`.

I see 3 options, let me know your thoughts @ceteri @Ankush-Chander on what option you prefer:

### Option 1: bring in the dependencies for everyone as part of base dependencies

`pandas` and `altair` are common dependencies IMO.

### Option 2: use extra dependenciese

In `setup.py`, I could add

```diff
        extras_require = {
            "base": base_packages,
            "docs": docs_packages,
+            "viz": ["altair"]
            },
```

And then users who want to use that feature can do `pip install pytextrank[viz]`

### Option 3: downgrade viz to matplotlib and bring that as a dependency

We could also write that method using matplotlib. The graph won't be as nice with the interactive tooltips etc... but the dependency wouldn't be as big.

```python
import matplotlib as mpl
from matplotlib import pyplot as plt

x = range(len(doc._.phrases))
height = [p.rank for p in doc._.phrases]
norm = mpl.colors.Normalize(vmin=0, vmax=max(p.count for p in doc._.phrases))
cmap = mpl.cm.ScalarMappable(norm=norm, cmap=mpl.cm.Blues)
cmap.set_array([])
plt.bar(x, height, width=0.8, color=[cmap.to_rgba(p.count) for p in doc._.phrases])
```

image

### PR DESCRIPTION
**Pull request by louisguitton**
fixes #158

## TODO

- [ ] check docs
- [ ] add image to Readme and docs ?

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -39,6 +39,7 @@ def parse_requirements_file (filename: str) -> typing.List:
 
     base_packages = parse_requirements_file("requirements.txt")
     docs_packages = parse_requirements_file("requirements-dev.txt")
+    viz_packages = parse_requirements_file("requirements-viz.txt")
```
Comment (louisguitton) [inline setup.py:42]: I would have added just 
```suggestion
    viz_packages = ["altair"]
```
but I saw the requirements-dev.txt file so I thought keeping it consistent was good.

What I do usually is that I parse the `requirements.txt` file for base packages (this way I use it in `setup.py` and in `Dockerfile` when docker is relevant) but for `dev`, `docs`, `viz`, ... I usually use arrays directly.

**Comment by louisguitton** [comment]
I built looked at the docs with 

```sh
mkdocs build --clean
python -m http.server -d site
open http://0.0.0.0:8000/
```
I'm missing the magic you do to turn the notebooks into `.md` files because right now I see
```txt
WARNING -  A relative path to 'sample.md' is included in the 'nav' configuration, which is not found in the documentation files
```

**Comment by ceteri** [comment]
Yes, the `sh bin/nb_md.sh` step is some obscure magic to run Selenium, `nbconvert`, and `bin/vis_doc.py` so that the example notebooks get rendered as Markdown. There were problems with the notebook plugin for MkDocs, plus some of the graphics use seemed to be difficult to render properly in the docs, so these scripts were workaround. Clumsy code, and maybe this can be rewritten at some point :)  Although probably the `mknotebooks` plugin improves first, so these are no longer needed.

There's another step `sh pkg_doc.py docs/ref.md` to regenerate the `docs/refs.md` apidocs. This depends on the `pyfixdoc.py` which I'd really like to move to its own library on PyPi.  It uses `inspect` since other apidocs generators seemed to have key points that they could not show very well, since the introduction of type annotations. Instead, this is a rewrite of those kinds of tool, which parses the code to help augment apidocs strings extracted from comments in the code. I was really trying to follow the Diataxis advice, where apidocs are less about coding samples, and more about *reference*, and in particular better at hyperlinks to definitive reference sources.  That said, almost every time we add new features to the code base here or in `kglab` my usage of `inspect` seems to break :)  It should probably not become a standalone library until this starts to stabilize.

I feel conflicted that these two steps deviate from typical practice with MkDocs for Python libraries -- it can become a risk to customize too much! These two points seem to add lots to the documentation, which I feel justifies the trade-offs. Does that seem reasonable?

I've tried to write some instructions here <https://derwen.ai/docs/ptr/build/#online-documentation> although these are perhaps simply "reminders" instead of good documentation. May need to rework that part.

**Review by ceteri — APPROVED** [review]
Wonderful, thank you  @louisguitton !

**Comment by ceteri** [comment]
@louisguitton, should `pandas` also become a dependency?

**Comment by ceteri** [comment]
Docs online are updated too <https://derwen.ai/docs/ptr/sample/#altair-visualisation>

**Comment by louisguitton** [comment]
> @louisguitton, should `pandas` also become a dependency?
> Although, with the `kglab` integration we might was well move `pandas` to a main dependency anyway?

`pandas` is a dependency of `altair` so for now it will be part of `pytextrank[viz]`. I did not include it explicitly because I'd rather delegate the dependency pinning to `altair` (which does it really well)

> Does that seem reasonable?
> I've tried to write some instructions here https://derwen.ai/docs/ptr/build/#online-documentation although these are perhaps simply "reminders" instead of good documentation. May need to rework that part.

@ceteri I now only realise the work that went in from you to make this work. I think it's reasonable if those issues with the out-of-the-box plugin were too annoying.

I myself use mkdocs notebook plugin here for example https://football-strava.guitton.co/ and I know exactly what you mean with the plots ... :/ here for example https://football-strava.guitton.co/notebooks/analysis/#Altair-Geoshape-plot , but I was too lazy to fix it.

I had missed the docs on building the docs. What I think would be the best way forward is just to write his as code you know? something as simple as a makefile like this

```makefile
build-docs:
  ./bin/nb_md.sh
  ./pkg_doc.py docs/ref.md
  mkdocs build

preview-docs:
  ./bin/preview.py
```

You could also go further and to Continuous Deployment using github actions if that saves you time (but you have the custom flask server deployment so maybe not relevant for you). If you used github pages, you could include the docs deployment in a separate github action workflow.

**Comment by ceteri** [comment]
Great! Yes, use a makefile is a great approach.

Deployment of docs involves a few step, probably not as readily implemented in GH Actions, although possibly...

1. `mkdocs build`
2. using `tar` to create an archive
3. upload to a Google Cloud Storage bucket
4. then on deployed server(s), retrieve latest release from bucket
5. distribute within the local director structure which Flask draws from

... then update Flask / Gunicorn / Nginx so that old stuff isn't stuck in cache :)

BTW, just had a **fantastic* conversation with Dan McCreary about their use of MkDocs to publish interoperable glossaries, where definitions can be federated. Here's a related article <https://dmccreary.medium.com/auto-generating-ekg-documents-2c85f1e45f44>   Dan would love to use the bibliography-from-RDF generator as a MkDocs plugin.  Meanwhile, Derwen has an API for *formative self-assessment*, i.e., defining self-paced quizzes that get rendered into Markdown -- and maybe Dan will take that as a MkDocs plugin.  In any case, I'm glad to see more widespread use of MkDocs and the community around its use in large projects, and better support for plugins.

**Pull request merged**